research-prototype/ipa/asort.py

```diff
@@ -0,0 +1,166 @@
+from Compiler.instructions import delshuffle
+from Compiler.library import for_range, break_point, if_, else_, if_e
+from Compiler.types import Array, Matrix, sint, regint
+from Compiler.sorting import radix_sort as mp_spdz_radix_sort
+
+
+def dest_comp(B):
+    """
+    Compute the 'destination' permutation
+
+    Calculate the permutation to stable sort a bit vector.
+    In the original, we multiply have of the cumulative sums
+    by the bits, and half by the complements of the bits.
+    This can be improved by just refactoring:
+    dest[i] = (1 - keys[i]) * cumval[i] + keys[i] * cumval[i + n]
+    = cumval[i] + keys[i] * (cumval[i + n] - cumval[i])
+    Note: this gives the destination for 1-origin indexing
+    for 0-origin (as in Python) we must subtract 1.
+    """
+    Bt = B.transpose()
+    Bt_flat = Bt.get_vector()
+    St_flat = Bt.value_type.Array(len(Bt_flat))
+    St_flat.assign(Bt_flat)
+    num = len(St_flat) // 2
+
+    @for_range(len(St_flat) - 1)
+    def _(i):
+        St_flat[i + 1] = St_flat[i + 1] + St_flat[i]
+    cumval = St_flat.get_vector(size=num)
+    cumshift = St_flat.get_vector(base=num, size=num) - cumval
+    dest = cumval + Bt_flat.get_vector(base=num, size=num) * cumshift - 1
+    Tt = Array(num, B.value_type)
+    Tt.assign_vector(dest)
+    return Tt
+
+
+def reveal_sort(k, D, reverse=False):
+    """
+    k is a permutation.
+    Rearrange D by k.
+    """
+    assert len(k) == len(D)
+    break_point()
+    shuffle = sint.get_secure_shuffle(len(k))
+    k_prime = k.get_vector().secure_permute(shuffle).reveal()
+    idx = Array.create_from(k_prime)
+    if D:
+        reverse.assign_vector(D.get_slice_vector(idx))
+        break_point()
+        D.secure_permute(shuffle, reverse=True)
+    else:
+        D.secure_permute(shuffle)
+        break_point()
+        v = D.get_vector()
+        D.assign_slice_vector(idx, v)
+    break_point()
+    delshuffle(shuffle)
+
+
+def double_dest(bs):
+    """
+    bs is an n by 2 bit array.
+    """
+    num, _ = bs.sizes
+    # BUG this is never used
+    bits = sint.Array(num * 4)
+    col0 = bs.get_column(0)
+    col1 = bs.get_column(1)
+    prod = col0 * col1
+    # BUG this is never assigned
+    cum.assign_vector(prod - col0 - col1 + 1)  # 00
```

Comment:
this is never assigned. should it be `bits`? if not, it needs to be renamed (this is a potentially offensive name.)